src/lib.rs

```diff
@@ -1,101 +1,405 @@
+    }
+    if a.range.start > b.range.start {
+        return Ordering::Greater;
+    }
+    if a.range.end < b.range.end {
+        return Ordering::Less;
+    }
+    if a.range.end > b.range.end {
+        return Ordering::Greater;
+    }
+    Ordering::Equal
+}
+
+fn has_overlapping_ranges(sorted_pending_fixes: &[PendingFix]) -> bool {
+    let mut prev_end = None;
+    for pending_fix in sorted_pending_fixes {
+        if let Some(prev_end) = prev_end {
+            if pending_fix.range.start < prev_end {
+                return true;
+            }
+        }
+        prev_end = Some(pending_fix.range.end);
+    }
+    false
+}
+
+fn get_sorted_non_conflicting_pending_fixes(
+    pending_fixes: HashMap<RuleName, Vec<PendingFix>>,
+) -> Vec<PendingFix> {
+    pending_fixes.into_iter().fold(
+        Default::default(),
+        |accumulated_fixes, (rule_name, mut pending_fixes_for_rule)| {
+            pending_fixes_for_rule.sort_by(compare_pending_fixes);
+            if has_overlapping_ranges(&pending_fixes_for_rule) {
+                panic!("Rule {:?} tried to apply self-conflicting fixes", rule_name);
+            }
+            let mut tentative = accumulated_fixes.clone();
+            tentative.extend(pending_fixes_for_rule);
+            if has_overlapping_ranges(&tentative) {
+                accumulated_fixes
+            } else {
+                tentative
+            }
         },
     )
-    .unwrap();
-    if reported_any_violations.load(Ordering::Relaxed) {
-        process::exit(1);
-    } else {
-        process::exit(0);
+}
+
+#[derive(Default)]
+struct AllPendingFixes(Mutex<HashMap<PathBuf, PerFilePendingFixes>>);
+
+impl AllPendingFixes {
+    pub fn append(
+        &self,
+        path: &Path,
+        file_contents: &[u8],
+        rule_name: &str,
+        fixes: Vec<PendingFix>,
+    ) {
+        self.lock()
+            .unwrap()
+            .entry(path.to_owned())
+            .or_insert_with(|| PerFilePendingFixes::new(file_contents.to_owned()))
+            .pending_fixes
+            .entry(rule_name.to_owned())
+            .or_default()
+            .extend(fixes);
+    }
+
+    pub fn into_inner(
+        self,
+    ) -> Result<
+        HashMap<PathBuf, PerFilePendingFixes>,
+        PoisonError<HashMap<PathBuf, PerFilePendingFixes>>,
+    > {
+        self.0.into_inner()
+    }
+}
+
+impl Deref for AllPendingFixes {
+    type Target = Mutex<HashMap<PathBuf, PerFilePendingFixes>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+struct PerFilePendingFixes {
+    file_contents: Vec<u8>,
+    pending_fixes: HashMap<RuleName, Vec<PendingFix>>,
+}
+
+impl PerFilePendingFixes {
+    fn new(file_contents: Vec<u8>) -> Self {
+        Self {
+            file_contents,
+            pending_fixes: Default::default(),
+        }
     }
 }
 
 type RuleIndex = usize;
 type RuleListenerIndex = usize;
 
-struct AggregatedQueries {
+struct AggregatedQueries<'resolved_rules> {
```

Comment:
In some places these "descriptive" lifetime names were not seeming that helpful and were seeming a little noisy so I was switching them to `'a`

Might feel that way here as well